### ISSUE TITLE
git_index_add_frombuffer creates 0160000 mode objects as blobs

### ISSUE THREAD

**Issue by ColMelvin** [libgit2/libgit2#4425]
The `git_index_add_frombuffer` function accepts the commit mode, `0160000`, when creating a new entry from a buffer.  However, it creates the commit mode entry as a blob.  Since a blob object and a commit object with the exact same data will have a different SHA-1 checksum, this behavior may lead to unexpected results.

Notably, `git_index_add_frombuffer` will return an error when the tree mode, `040000`, is provided for the mode.  I think the same should happen for `0160000`.

### Reproduction steps
The reproduction code below uses a modified version of Perl's Git::Raw v0.74, available at https://github.com/jacquesg/p5-Git-Raw/pull/191.  This modification will hopefully be included in v0.75.

```perl
#!/usr/bin/perl

use strict;
use warnings;

use File::Temp;
use Git::Raw;
use Test::More tests => 3;

my $dir   = File::Temp->newdir();
my $repo  = Git::Raw::Repository->init( $dir->dirname, 1 );
my $index = $repo->index;

my $COMMIT_DATA = "tree 4b825dc642cb6eb9a060e54bf8d69288fbee4904\x0aauthor A U Thor <author\@example.com> 1234567890 +0000\x0acommitter C O Mitter <committer\@example.com> 1234567890 +0000\x0a\x0aEmpty\x0a";
my $COMMIT_SHA  = '746262f5c8ed0115432581c2bcf8f08c77ae76c8';

$index->add_frombuffer( 'submodule', $COMMIT_DATA, 0160000 );
my $tree = $index->write_tree;

my ($entry) = $tree->entries;
is $entry->file_mode,  0160000;               # Passes
isa_ok $entry->object, 'Git::Raw::Commit';    # Fails
is $entry->object->id, $COMMIT_SHA;           # Fails

exit;

# Sanity check proving actual commit value
my $hash = qx{cd '$dir'; echo -n '$COMMIT_DATA' | git hash-object -t commit --stdin};
chomp $hash;
is $hash, $COMMIT_SHA;                        # Passes
```

### Expected behavior

Either:
- all tests pass, or
- script dies at `$index->add_frombuffer( 'submodule', $COMMIT_DATA, 0160000 );` with 'invalid filemode'

### Actual behavior

Script runs to completion; 2nd & 3rd tests fail.

```
1..3
ok 1
not ok 2 - An object of class 'Git::Raw::Blob' isa 'Git::Raw::Commit'
#   Failed test 'An object of class 'Git::Raw::Blob' isa 'Git::Raw::Commit''
#   at proof line 22.
#     The object of class 'Git::Raw::Blob' isn't a 'Git::Raw::Commit'
not ok 3
#   Failed test at proof line 23.
#          got: 'f8aabe3d076ed2278a652ba6d76ea152148daa3c'
#     expected: '746262f5c8ed0115432581c2bcf8f08c77ae76c8'
# Looks like you failed 2 tests of 3.
```

### Version of libgit2 (release number or SHA1)
Tested Git::Raw v0.74 (ee89941fa2 of libgit2, per Changes file).

Inspection of the code suggests the issue remains (as of 364184acca0d993deaeb060197b595f7c49ea226)

### Operating system(s) tested
Linux 3.10.0-514.26.2.el7.x86_64

**Comment by ethomson** [comment]
I agree.  This seems like a pretty straightforward - do you have an interest in creating a pull request?  😀

**Comment by ColMelvin** [comment]
I can probably get to it in a couple days

**Comment by ethomson** [comment]
Opened a pull request to fix this; see #4529
